### DISCUSSION TITLE
Bug? TypeError: headers is undefined (user-agent.js (43:1))

### DISCUSSION THREAD

**Discussion by kvirulent** [vercel/next.js#67933] [Help] [answered]
### Summary

I'm trying to use userAgent to detect if the client is a mobile device, and redirect them to another route. However, Nextjs returns an error that points at an issue in the user-agent module:

TypeError: headers is undefined
node_modules\next\dist\server\web\spec-extension\user-agent.js (43:1)

Additionally this error is present on the server console:
TypeError: Cannot read properties of undefined (reading 'get')
    at Home (./app/page.tsx:20:72)
digest: "1631309234"
 GET / 500 in 99ms

I'm not sure if I've done something wrong or this is an issue with userAgent, but before creating an issue I thought I'd ask first. A devtunnel to the project is linked.

### Additional information

```js
// ./app/page.tsx:

"use client";
import Startup from "@/app/ui/startup";
import Link from "next/link";
import CubeAnimation from "@/app/ui/cube2";
import { NextRequest, userAgent } from 'next/server';
import { useRouter } from 'next/navigation';

export default function Home(request: NextRequest) {
  const router = useRouter();
  const user = userAgent(request);
  if (user.device.type == "mobile") {
    router.push("/index-nostartup")
  };

  return (
    <main className="min-h-screen text-amber glow">
      {/* Startup Content */}
      <div>
        <Startup />
      </div>

      {/* Page Content */}
      <div className="flex items-center justify-center">
        <div className="text-amber-dim md:text-amber max-md:w-fit max-md:mx-auto max-md:absolute md:flex-1 h-screen flex items-center justify-center">
          <div className="fixed z-[1] cursor-default">
            <CubeAnimation size={190} />
          </div>
        </div>
        <div className="md:flex-1 h-screen flex flex-col z-[10] items-center justify-center">
          <h1 className="text-4xl md:text-6xl m-5">xirulent.dev</h1>
          <ul>
            <li className="m-5 w-fit hover:box-glow hover:bg-[#FFB000] hover:text-black">
              <Link href="/pages/about-me">&gt; who r u?</Link>
            </li>
            <li className="m-5 w-fit hover:box-glow hover:bg-[#FFB000] hover:text-black">
              <Link href="#">&gt; where am i?</Link>
            </li>
            <li className="m-5 w-fit hover:box-glow hover:bg-[#FFB000] hover:text-black">
              <Link href="#">&gt; option_3</Link>
            </li>
            <li className="m-5 w-fit hover:box-glow hover:bg-[#FFB000] hover:text-black">
              <Link href="/pages/webring">&gt; webring</Link>
            </li>
          </ul>
        </div>
      </div>
    </main>
  );
}
```

### Example

https://qhxdl49k-3000.usw3.devtunnels.ms

**Comment by icyJoseph** [comment, marked as answer]
Alright, so there's a concepts misunderstanding here.

Both Server Component Pages, and Client Component Pages, receive only, params and searchParams as props. These are not called, ever, with the request object. I think maybe you are confusing the Route Handlers, with the Page entry points. 

Something like:

```tsx
// app/blog/[slug]/page.tsx
export default function Page({
  params,
  searchParams,
}: {
  params: { slug: string }
  searchParams: { [key: string]: string | string[] | undefined }
}) {
  return <h1>My Page</h1>
}
```

And then there's the issue that you are calling `router.push` as you render, that's a side-effect, and render passes should not have side-effects.

I'd recommend:

1. remove the `use client` directive
2. import redirect from next/navigation: `import { redirect } from 'next/navigation'`
3. use the headers function get the user agent: `import { headers } from 'next/headers'`
4. call redirect, as you render (like you have it):
```tsx
  if (/* you need to check for mobile here*/) {
    redirect("/index-nostartup")
  };
```

Since the `userAgent` function, only cares about headers, as per, https://github.com/vercel/next.js/blob/76b9e52aff1823e97e68094bb3d709a5fc8d8c54/packages/next/src/server/web/spec-extension/user-agent.ts#L42

```tsx
export function userAgent({ headers }: { headers: Headers }): UserAgent {
  return userAgentFromString(headers.get('user-agent') || undefined)
}
```

You can pass the headers you get from calling the headers function, like: `userAgent({ headers: headers() })`

**Reply by kvirulent** [reply]
Thanks, I realized this solution is way over engineered when I could just use CSS to hide the startup component.
